### ISSUE TITLE
vcluster plugin hooks

### ISSUE THREAD

**Issue by FabianKramm** [loft-sh/vcluster#465]
Currently vcluster lets you write plugin and create completely separate controllers and syncers with the plugin SDK. However there are many use cases that do not want to exchange a syncer or create a new one from scratch and rather want to change syncing behaviour for existing syncers. We should introduce hooks that allows plugins to modify outgoing objects similar to an admission controller.

**Comment by olljanat** [comment]
This would be very useful for my [RunAsUser Plugin](https://github.com/k8s4u/vcluster-runasuser-plugin) which currently uses plugin functionality to just register it selves as *MutatingAdmissionWebhook* and targets to make users life easier with restricted pods.

However it looks to be that currently I cannot get rid of `PodSecurity` warnings coming from vcluster because validation actually happens before *MutatingAdmissionWebhook* modify pod spec.

Here output from test case. As you can see there is warning about `runAsNonRoot` and pod still gets started fine and run as non-root:
```
$ vcluster create my-vcluster -n my-vcluster -f https://raw.githubusercontent.com/k8s4u/vcluster-runasuser-plugin/main/plugin.yaml
[info]   Creating namespace my-vcluster
[info]   execute command: helm upgrade my-vcluster vcluster --repo https://charts.loft.sh --version 0.8.0 --kubeconfig /tmp/2563261166 --namespace my-vcluster --install --repository-config='' --values /tmp/222495219 --values https://raw.githubusercontent.com/k8s4u/vcluster-runasuser-plugin/main/plugin.yaml
[done] √ Successfully created virtual cluster my-vcluster in namespace my-vcluster.
- Use 'vcluster connect my-vcluster --namespace my-vcluster' to access the virtual cluster
- Use `vcluster connect my-vcluster --namespace my-vcluster -- kubectl get ns` to run a command directly within the vcluster

$ vcluster connect my-vcluster --namespace my-vcluster -- kubectl create deployment test --image=busybox -- sleep infinity
[info]   Waiting for vcluster to come up...
Warning: would violate PodSecurity "restricted:latest": allowPrivilegeEscalation != false (container "busybox" must set securityContext.allowPrivilegeEscalation=false), unrestricted capabilities (container "busybox" must set securityContext.capabilities.drop=["ALL"]), runAsNonRoot != true (pod or container "busybox" must set securityContext.runAsNonRoot=true), seccompProfile (pod or container "busybox" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")
deployment.apps/test created

$ vcluster connect my-vcluster --namespace my-vcluster -- kubectl get pods
NAME                    READY   STATUS    RESTARTS   AGE
test-647d7cd798-pxlkd   1/1     Running   0          6s

$ vcluster connect my-vcluster --namespace my-vcluster -- kubectl exec -it test-647d7cd798-pxlkd -- whoami
whoami: unknown uid 1000126939
command terminated with exit code 1
```

**Comment by Shaked** [comment]
@olljanat  I share the same thoughts as you about this issue. This would be a great feature in vcluster. 

In the meanwhile, I solved this by using jsoolicy - another tool by Loft. I didn’t want to code the admission hook myself, so I ended up with this policy: 

https://github.com/Greeneye-Technology/vcluster-secrets-store-csi-plugin/blob/main/jspolicy/secret-provider-class.yaml

It requires the user to install jspolicy and the policy itself on the host.
